File: backend/app/schemas/social.py
"""API contracts for friendships and private/global regularity competitions."""

from __future__ import annotations

import uuid
from datetime import date
from typing import Literal

from pydantic import BaseModel, Field, model_validator

CompetitionMetric = Literal[
    "regularity",
    "weight_loss",
    "waist_reduction",
    "relative_strength",
]


class CompetitionFactor(BaseModel):
    metric: CompetitionMetric
    exercise_id: uuid.UUID | None = None

    @model_validator(mode="after")
    def validate_exercise(self) -> "CompetitionFactor":
        if self.metric == "relative_strength" and self.exercise_id is None:
            raise ValueError("Для силового прогресса выберите упражнение")
        if self.metric != "relative_strength" and self.exercise_id is not None:
            raise ValueError("Упражнение можно выбрать только для силового прогресса")
        return self

    @property
    def key(self) -> str:
        if self.exercise_id is None:
            return self.metric
        return f"{self.metric}:{self.exercise_id}"


class CompetitionFactorSummary(BaseModel):
    key: str
    metric: CompetitionMetric
    label: str
    exercise_id: uuid.UUID | None = None


class CompetitionFactorResult(BaseModel):
    key: str
    metric: CompetitionMetric
    label: str
    status: Literal["ready", "baseline_missing", "no_progress"]
    value: float | None = None
    completed: int | None = None
    planned: int | None = None
    baseline_value: float | None = None
    latest_value: float | None = None
    baseline_date: date | None = None
    latest_date: date | None = None
    unit: str | None = None
    capped: bool = False


class CompetitionParticipantAnalytics(BaseModel):
    wins: int = 0
    factors: list[CompetitionFactorResult]


class FriendSummary(BaseModel):
    id: uuid.UUID
    label: str
    status: Literal["accepted", "blocked"]


class FriendListResponse(BaseModel):
    items: list[FriendSummary]


class CompetitionScoreResponse(BaseModel):
    score: float | None = None
    completed: int
    planned: int


class CompetitionSummary(BaseModel):
    id: uuid.UUID
    friendship_id: uuid.UUID
    friend_label: str
    status: Literal["pending", "active", "finished", "cancelled"]
    title: str | None = None
    duration_days: int = Field(ge=7, le=365)
    start_date: date | None = None
    end_date: date | None = None
    algorithm_version: str
    created_by_me: bool
    can_accept: bool
    factors: list[CompetitionFactorSummary] = Field(default_factory=list)
    winner: Literal["me", "friend", "tie"] | None = None
    my_analytics: CompetitionParticipantAnalytics | None = None
    friend_analytics: CompetitionParticipantAnalytics | None = None
    my_score: CompetitionScoreResponse | None = None
    friend_score: CompetitionScoreResponse | None = None


class CompetitionListResponse(BaseModel):
    items: list[CompetitionSummary]


class CompetitionCreateRequest(BaseModel):
    friendship_id: uuid.UUID
    title: str | None = Field(default=None, max_length=120)
    duration_days: int = Field(default=14, ge=7, le=365)
    factors: list[CompetitionFactor] = Field(
        default_factory=lambda: [CompetitionFactor(metric="regularity")],
        min_length=1,
        max_length=4,
    )

    @model_validator(mode="after")
    def unique_factors(self) -> "CompetitionCreateRequest":
        keys = [factor.key for factor in self.factors]
        if len(keys) != len(set(keys)):
            raise ValueError("Факторы соревнования не должны повторяться")
        if self.title is not None:
            self.title = self.title.strip() or None
        return self


class SocialActionResponse(BaseModel):
    ok: bool = True


class GlobalLeaderboardEntry(BaseModel):
    rank: int
    alias: str
    score: float
    completed: int
    planned: int
    is_me: bool = False


class GlobalSeasonResponse(BaseModel):
    season_key: str
    title: str
    start_date: date
    end_date: date
    join_deadline: date
    status: Literal["not_joined", "joined", "left"]
    algorithm_version: str
    cohort: Literal["days_1_2", "days_3", "days_4_plus"]
    cohort_label: str
    participant_count: int
    minimum_cohort_size: int
    ranking_unlocked: bool
    ranked_eligible: bool
    provisional: bool
    my_alias: str | None = None
    my_rank: int | None = None
    my_score: CompetitionScoreResponse | None = None
    leaderboard: list[GlobalLeaderboardEntry]

File: frontend/src/api/social.ts
import { z } from "zod";

import { apiClient } from "@/api/client";

const friendSchema = z.object({
  id: z.string().uuid(),
  label: z.string().min(1),
  status: z.enum(["accepted", "blocked"]),
});

const scoreSchema = z.object({
  score: z.number().nullable(),
  completed: z.number().int().nonnegative(),
  planned: z.number().int().nonnegative(),
});

const competitionMetricSchema = z.enum([
  "regularity",
  "weight_loss",
  "waist_reduction",
  "relative_strength",
]);

const factorSchema = z.object({
  key: z.string().min(1),
  metric: competitionMetricSchema,
  label: z.string().min(1),
  exercise_id: z.string().uuid().nullable().optional(),
});

const factorResultSchema = factorSchema.pick({ key: true, metric: true, label: true }).extend({
  status: z.enum(["ready", "baseline_missing", "no_progress"]),
  value: z.number().nullable(),
  completed: z.number().int().nonnegative().nullable().optional(),
  planned: z.number().int().nonnegative().nullable().optional(),
  baseline_value: z.number().nullable().optional(),
  latest_value: z.number().nullable().optional(),
  baseline_date: z.string().nullable().optional(),
  latest_date: z.string().nullable().optional(),
  unit: z.string().nullable().optional(),
  capped: z.boolean().default(false),
});

const participantAnalyticsSchema = z.object({
  wins: z.number().int().nonnegative(),
  factors: z.array(factorResultSchema),
});

const competitionSchema = z.object({
  id: z.string().uuid(),
  friendship_id: z.string().uuid(),
  friend_label: z.string().min(1),
  status: z.enum(["pending", "active", "finished", "cancelled"]),
  title: z.string().nullable().optional(),
  duration_days: z.number().int().min(7).max(365),
  start_date: z.string().nullable(),
  end_date: z.string().nullable(),
  algorithm_version: z.string(),
  created_by_me: z.boolean(),
  can_accept: z.boolean(),
  factors: z.array(factorSchema).default([]),
  winner: z.enum(["me", "friend", "tie"]).nullable().optional(),
  my_analytics: participantAnalyticsSchema.nullable().optional(),
  friend_analytics: participantAnalyticsSchema.nullable().optional(),
  my_score: scoreSchema.nullable(),
  friend_score: scoreSchema.nullable(),
});

const globalLeaderboardEntrySchema = z.object({
  rank: z.number().int().positive(),
  alias: z.string().min(1),
  score: z.number().min(0).max(100),
  completed: z.number().int().nonnegative(),
  planned: z.number().int().nonnegative(),
  is_me: z.boolean(),
});

const globalSeasonSchema = z.object({
  season_key: z.string().min(1),
  title: z.string().min(1),
  start_date: z.string(),
  end_date: z.string(),
  join_deadline: z.string(),
  status: z.enum(["not_joined", "joined", "left"]),
  algorithm_version: z.string().min(1),
  cohort: z.enum(["days_1_2", "days_3", "days_4_plus"]),
  cohort_label: z.string().min(1),
  participant_count: z.number().int().nonnegative(),
  minimum_cohort_size: z.number().int().positive(),
  ranking_unlocked: z.boolean(),
  ranked_eligible: z.boolean(),
  provisional: z.boolean(),
  my_alias: z.string().nullable(),
  my_rank: z.number().int().positive().nullable(),
  my_score: scoreSchema.nullable(),
  leaderboard: z.array(globalLeaderboardEntrySchema),
});

export type Friend = z.infer<typeof friendSchema>;
export type Competition = z.infer<typeof competitionSchema>;
export type CompetitionMetric = z.infer<typeof competitionMetricSchema>;
export type CompetitionFactorInput = {
  metric: CompetitionMetric;
  exercise_id?: string;
};
export type GlobalSeason = z.infer<typeof globalSeasonSchema>;

export async function getFriends(): Promise<Friend[]> {
  const { data } = await apiClient.get("/friends");
  return z.object({ items: z.array(friendSchema) }).parse(data).items;
}

export async function getCompetitions(): Promise<Competition[]> {
  const { data } = await apiClient.get("/competitions");
  return z.object({ items: z.array(competitionSchema) }).parse(data).items;
}

export async function createFriendCompetition(
  friendshipId: string,
  durationDays: number,
  factors?: CompetitionFactorInput[],
  title?: string,
): Promise<void> {
  await apiClient.post("/competitions/friend", {
    friendship_id: friendshipId,
    duration_days: durationDays,
    ...(factors ? { factors } : {}),
    ...(title ? { title } : {}),
  });
}

export async function acceptCompetition(competitionId: string): Promise<void> {
  await apiClient.post(`/competitions/${competitionId}/accept`);
}

export async function leaveCompetition(competitionId: string): Promise<void> {
  await apiClient.post(`/competitions/${competitionId}/leave`);
}

export async function changeFriendship(
  friendshipId: string,
  action: "remove" | "block" | "unblock",
): Promise<void> {
  await apiClient.post(`/friends/${friendshipId}/${action}`);
}

export async function getGlobalSeason(): Promise<GlobalSeason> {
  const { data } = await apiClient.get("/competitions/global/current");
  return globalSeasonSchema.parse(data);
}

export async function joinGlobalSeason(): Promise<void> {
  await apiClient.post("/competitions/global/current/join");
}

export async function leaveGlobalSeason(): Promise<void> {
  await apiClient.post("/competitions/global/current/leave");
}
